FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc_startup.s

```diff
@@ -0,0 +1,265 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+--------------------------------------------- */
+
+/**
+ * @file
+ * @ingroup ARC_HAL_STARTUP
+ * @brief assembly part of startup process
+ */
+
+/**
+ * @addtogroup ARC_HAL_STARTUP
+ * @{
+ */
+/** @cond STARTUP_ASM */
+
+#define __ASSEMBLY__
+#include "arc/arc.h"
+#include "arc/arc_asm_common.h"
+
+	.file "arc_startup.s"
+
+#ifdef __GNU__
+.weak	__SDATA_BEGIN__		/* start of small data, defined in link script */
+#else
+.weak	_f_sdata		/* start of small data, defined in link script */
+#endif
+.weak	init_hardware_hook	/* app hardware init hook */
+
+.extern	board_main
+.extern exc_entry_table
+
+/* initial vector table */
+	.section .init_vector, "a"
+	.long _arc_reset
+	.section .init_bootstrap, "ax"
+	.global _arc_reset
+	.global _start
+	.align 4
+_start:
+_arc_reset:
+_arc_reset_stage1:
+	kflag  STATUS32_RESET_VALUE
+
+/* STAGE 1 */
+
+/* necessary hardware should be done first to speed up initialization
+	1. system clk
+	2. mem controller must be initialized before any access to external
+	mem.
+	3. others
+*/
+_arc_cache_init_start:
+	lr	r0, [AUX_BCR_D_CACHE]
+	cmp	r0, 2
+	/* invalidate dcache */
+	jle	_arc_icache_init
+	mov	r0, 1
+	sr	r0, [AUX_DC_IVDC]
+_arc_icache_init:
+	lr	r0, [AUX_BCR_I_CACHE]
+	cmp	r0, 2
+	jle	_arc_cache_init_end
+	/* invalidate icache */
+	mov	r0, 1
+	sr	r0, [AUX_IC_IVIC]
+	nop_s
+	nop_s
+	nop_s
+	sr	r0, [AUX_IC_CTRL]
+
+_arc_cache_init_end:
+	mov	r0, init_hardware_hook
+	cmp	r0, 0
+	jlne	[r0]
+
+/* STAGE 2: init necessary registers */
+
+_arc_reset_stage2:
+	mov	r0, 0
+
+/* interrupt related init */
+	sr	r0, [AUX_IRQ_ACT]
+	sr	r0, [AUX_IRQ_CTRL]
+	sr	r0, [AUX_IRQ_HINT]
+
+/* use the new vector table to replace the old one */
+#if defined(ARC_FEATURE_SEC_PRESENT) && (SECURESHIELD_VERSION < 2)
+	sr	exc_entry_table, [AUX_INT_VECT_BASE_S]
+#else
+	sr	exc_entry_table, [AUX_INT_VECT_BASE]
+#endif
+#ifdef __GNU__
+	mov	gp, __SDATA_BEGIN__	/* init small-data base register */
+#else
+	mov	gp, _f_sdata	/* init small-data base register */
+#endif
+	mov	fp, 0		/* init fp register */
+
+#if ARC_FEATURE_MP_NUM_CPUS > 1
+	GET_CORE_ID r0
+	breq r0, 0, _master_core_startup
+
+_slave_core_wait:
+	ld r1, [arc_cpu_wake_flag]
+	brne r0, r1, _slave_core_wait
+
+	ld sp, [arc_cpu_sp]
+	/* signal master core that slave core runs */
+	st 0, [arc_cpu_wake_flag]
+
+	j arc_slave_start
+
+_master_core_startup:
```

Comment:
We don't support SMP in this sample - let's drop everything related to it here (and in other places)